dietpi/dietpi-software

```diff
@@ -4571,6 +4579,102 @@ amvdec_vp9' > /etc/modules-load.d/dietpi-c4-kodi.conf
 
 		fi
 
+		software_id=171 # frp
+		if (( ${aSOFTWARE_INSTALL_STATE[$software_id]} == 1 )); then
+
+			Banner_Installing
+
+			local arch='arm'
+
+			# - ARMv8
+			if (( $G_HW_ARCH == 3 )); then
+
+				arch='arm64'
+
+			# - x86_64
+			elif (( $G_HW_ARCH == 10 )); then
+
+				arch='amd64'
+
+			fi
+
+			local fallback_url="https://github.com/fatedier/frp/releases/download/v0.36.2/frp_0.36.2_linux_$arch.tar.gz"
+			local pattern="\"browser_download_url\": .*\/frp_[0-9.]*_linux_$arch\.tar\.gz\""
+			Download_Install "$(curl -sSfL 'https://api.github.com/repos/fatedier/frp/releases/latest' | mawk -v pat="$pattern" -F\" '$0 ~ pat {print $4}')"
+
+			G_EXEC cd frp_*
+
+			local choice_required=
+			while :
+			do
+
+				G_WHIP_MENU_ARRAY=(
+					'Server' ': Use this machine as a server, with a public IP'
+					'Client' ': Use this machine as a client, without a public IP'
+					'Both' ': Run the reverse proxy only on this machine'
+				)
+
+				G_WHIP_MENU "${choice_required}Please choose how you are going to run frp." && break
+				choice_required='[ERROR] A choice is required to finish the frp install.\n\n'
+
+			done
+
+			G_EXEC mkdir /etc/frp
+
+			if [[ $G_WHIP_RETURNED_VALUE == 'Server' || $G_WHIP_RETURNED_VALUE == 'Both' ]]; then
+				G_EXEC mv frps /usr/bin/frps
+				G_EXEC mv systemd/frps.service /etc/systemd/system/frps.service
+
+				# Pre-create config file to turn on dashboard
+				cat << _EOF_ > /etc/frp/frps.ini
+[common]
+bind_port = 7000
+
+dashboard_port = 7500
+dashboard_user = admin
+dashboard_pwd = $GLOBAL_PW
```

Comment:
Is it possible to use basic HTTP authentication here or so, at least storing credentials/password in a dedicated file that can be better secured? I try to not store the global software password in plain text, where possible, at least not world-readable. If all does not work, at least we can change the config file permissions so that it's writable only by root, readable only be frp and others have no read permissions either.

Reply:
It actually does use Basic Auth, but I think we'll need to make the file only root (and frp) readable.

Reply:
Ah, yes I meant to have the password stored in a hashed way like in `.htpasswd`.

Reply:
Maybe I should clarify: While it does let you authenticate using Basic Authentication, it hashes and encodes the password itself.  I don't quite get what you're saying.

Reply:
Let me try to clarify:
- With the current implementation, an evil user could do `cat /etc/frp/frps.ini` to see the global software password in plain text, which is the default for many other software titles and, if not changed by the user, even the root and dietpi UNIX users login passwords. This should be prevented, when possible.
- Best would be to store the password in `/etc/frp/frps.ini` directly in a hashed way. So that the config only contains the hash while the actual password is not stored anywhere. Many software titles support this, but not all.
- Second best option would be to store it (probably together with the `dashboard_user`) in a separate authentication file, which can be specifically hardened. But I guess that is not possible, also it makes things more complicated.
- Third best option is to harden permissions of `/etc/frp/frps.ini` itself: `chmod 0600 /etc/frp/frps.ini`
    This limit read access to the service user (currently root) only, which can be seen as downside, when it's about reviewing less sensitiv config values.

Reply:
Ah, just found https://github.com/fatedier/frp#split-configures-into-different-files (how did I miss that?).  Turns out you can split it into separate files!

Reply:
PR up: https://github.com/fatedier/frp/pull/2448.